README.md
# DependenTree

A graph visualization library built on top of D3. Displays a directed graph in a collapsible tree format. Users can view information about graph nodes (entities) and their dependencies.

## Table of Contents

- [DependenTree](#dependentree)
  - [Table of Contents](#table-of-contents)
  - [Features](#features)
  - [Live Code Playground](#live-code-playground)
  - [Quickstart](#quickstart)
  - [Terminology](#terminology)
  - [API](#api)
    - [Methods](#methods)
    - [Method Arguments](#method-arguments)
    - [Options](#options)
      - [Tree Behavior](#tree-behavior)
      - [Tree Appearance](#tree-appearance)
    - [Validate](#validate)
  - [Input Data Format](#input-data-format)
    - [Array Example](#array-example)
    - [Object Example](#object-example)
    - [Entity Object Format](#entity-object-format)
      - [Fields](#fields)
      - [Additional Fields](#additional-fields)
      - [Added Fields](#added-fields)
  - [Understanding Max Depth](#understanding-max-depth)
    - [Hypothetical Example](#hypothetical-example)
    - [Real Example](#real-example)
    - [Another Hypothetical](#another-hypothetical)
  - [Examples](#examples)
    - [`modifyEntityName`](#modifyentityname)
    - [`textClick`](#textclick)
    - [Form Example](#form-example)
  - [Resources](#resources)

## Features

- Visualize large dependency graphs, tested up to 14K nodes
- Supports visualization of cyclic graphs
- See both upstream and downstream dependencies
- Handle all graph mapping logic on the client-side
- Only requires a minimal input of node and edge data
- Simple API, no need to write D3 code

## Live Code Playground

Start with this [interactive walkthrough](https://observablehq.com/@amogh/dependentree) to learn about the library and its features.

## Quickstart

**DependenTree** takes graph data in JSON format. Construct your graph with the following API

```json
[
 {
    "_name": "Elizabeth II",
    "Title": "Queen"
  },
  {
    "_name": "Phillip",
    "Title": "Duke of Edinburgh"
  },
  {
    "_name": "Charles",
    "_deps": ["Phillip", "Elizabeth II"],
    "Title": "Prince of Wales"
  },
  {
    "_name": "Diana",
    "Title": "Princess of Wales"
  },
  {
    "_name": "William",
    "_deps": ["Diana", "Charles"],
    "Title": "Prince, Duke of Cambridge"
  },
  {
    "_name": "Catherine",
    "Title": "Duchess of Cambridge"
  },
  {
    "_name": "George",
    "_deps": ["Catherine", "William"]
  },
  {
    "_name": "Charlotte",
    "_deps": ["Catherine", "William"]
  },
  {
    "_name": "Louis",
    "_deps": ["Catherine", "William"]
  }
]
```

Above is a small graph of nine entities; in this case the data is people in a family tree. You'll note that each of the  children have their parents as dependencies. These dependencies are represented as an array of strings under the `_deps` key. These strings exactly match the `_name` of another entity. When this data is displayed Their `_name` will be displayed on the tree. Their `Title` will be displayed on a tooltip when hovering over each entity node.

```html
<html>
  <head>
    <style>
      * {
        padding: 0;
        margin: 0;
        box-sizing: border-box;
      }

      body,
      html {
        height: 100vh;
        width: 100%;
        overflow: auto;
      }
    </style>
  </head>
  <body><div id="tree"></div></body>
  <script type="text/javascript" src="dependentree.js"></script>
  <!-- For this example, the JSON array is parsed and saved as a JavaScript variable called royals -->
  <script type="text/javascript" src="royals.js"></script>
  <script>
    // Creates an instance of the class
    // The tree attaches to the body element
    const tree = new DependenTree('div#tree');

    // Adds dependency data to the tree
    tree.addEntities(royals);

    // Sets the entity on the tree, displays the upstream dependencies
    // You can also pass 'downstream' to display downstream dependencies
    tree.setTree('Elizabeth II', 'upstream');
  </script>
</html>
```

DependenTree is exported in the universal module definition format.

Install with  npm or yarn

```sh
npm install @square/dependentree

yarn add @square/dependentree
```

Import the `DependenTree` class

It is best to use `import` statements and allow your application to bundle the code from `src/index.js`. Using `require()` statements will pull from `/dist/dependentree.js` which comes already bundled.

```javascript
import DependenTree from '@square/dependentree';

const DependenTree = require('@square/dependentree');
```

```javascript
import royals from './testData';

// Creates an instance of the class
// The tree attaches to our designated element
const tree = new DependenTree('div#tree');

// Adds dependency data to the tree
tree.addEntities(royals);

// Sets the entity on the tree, displays the upstream dependencies
// You can also pass 'downstream' to display downstream dependencies
tree.setTree('Elizabeth II', 'upstream');
```

## Terminology

| Term        |                                                                                                                                                      |
| ----------- | ---------------------------------------------------------------------------------------------------------------------------------------------------- |
| Entity      | Entities are nodes in a graph. If our graph represents a family tree, an entity would be an individual person. Entities are unique to the graph.     |
| Entity Data | Refers to the JavaScript object that represents the information about the entity including it's name and list of dependencies.                       |
| Graph       | The graph which represents the dependency relationships. The graph exists in memory. A portion of data from the graph is pulled to create each tree. |
| Tree        | The visual tree structure on the page.                                                                                                               |
| Node        | An individual node on the visual tree structure. Nodes consist of text and a circular button. Each entity can have multiple nodes on a page.         |


## API

### Methods

| Method          | Example                                           | Description                                                                                                                                                                                                                                                         |
| --------------- | ------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `constructor`   | `new DependenTree('div#tree', {maxDepth: 25});`   | Creates an instance of the DependenTree class. It attaches to an HTML element on the page and passes [tree options](#options) to change the tree's appearance.                                                                                                      |
| `addEntities`   | `[{ "_name": "Elizabeth II", "Title": "Queen" }]` | List of entities and their dependencies. See the [Input Data Format](#input-data-format) section below.                                                                                                                                                             |
| `setTree`       | `.setTree('Elizabeth II', 'downstream');`         | Sets a new tree diagram to be displayed in the target HTML element. Also removes any previous tree in the target element.                                                                                                                                           |
| `removeTree`    | `.removeTree();`                                  | Removes the tree currently being displayed.                                                                                                                                                                                                                         |
| `expandAll`     | `.expandAll(4);`                                  | Opens all nodes in the tree up to the target depth represented as an integer. If no target depth is passed in, all nodes will be expanded. This method is not recommended for large trees. You are likely to experience layout thrashing and possibly a page crash. |
| `collapseAll`   | `.collapseAll();`                                 | Collapses all nodes. If `true` is passed in, the second level remains open along with the root node.                                                                                                                                                                |
| `getEntityList` | `.getEntityList('Title', 'Prince');`              | Returns a list of entities based on a `key: value` filter. If no filter is passed in, all entities will be returned.                                                                                                                                                |
| `validate`      | `.validate();`                                    | After passing data to `addEntities`, `validate` provides information about the quality of your input data. See [validate](#validate) section below.                                                                                                                 |

### Method Arguments

| Method          | Argument        | Type            | Required | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| --------------- | --------------- | --------------- | -------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `constructor`   | Selector String | String          | Yes      | [W3C Selector String](https://www.w3.org/TR/selectors-api/) for selecting your container element. This element is given a style of `overflow:auto;`. The tree will be appended to this element. Be sure to define this element's width before calling DependenTree. It is also best not to use the body element. An element that is a descendant of body works best. If you use the body, there may be issues with the automatic scrolling feature when switching between upstream and downstream. |
|                 | Options         | Object          | No       | Options to modify the graph appearance. See the [Options documentation](#options) below.                                                                                                                                                                                                                                                                                                                                                                                                           |
| `addEntities`   | Dependencies    | Object or Array | Yes      | List of entities and their dependencies. See the [Input Data Format](#input-data-format) section below.                                                                                                                                                                                                                                                                                                                                                                                            |
| `setTree`       | Entity Name     | String          | Yes      | The entity that will be the root node of the dependency tree.displayed.                                                                                                                                                                                                                                                                                                                                                                                                                            |
|                 | Direction       | String          | No       | This string sets whether the `'upstream'` or `'downstream'` dependencies of a given entity are displayed. This argument defaults to `'upstream'`.                                                                                                                                                                                                                                                                                                                                                  |
| `removeTree`    | n/a             | n/a             | No       | n/a                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| `expandAll`     | Target Depth    | Integer         | No       | This field expands all the nodes to a given depth. If no depth argument is passed in, all nodes in the tree will be expanded.                                                                                                                                                                                                                                                                                                                                                                      |
| `collapseAll`   | Second Level    | Boolean         | No       | If `true` is passed in, the second level will remain open along with the first.                                                                                                                                                                                                                                                                                                                                                                                                                    |
| `getEntityList` | Key             | String          | No       | The `key` of which to filter by. E.g. `'Title'`.                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|                 | Value           | String          | No       | The `value` of which to filter by. E.g. `'Queen'`.                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| `validate`      | n/a             | n/a             | No       | n/a                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |

### Options

In the constructor, DependenTree takes an options object. The various options control how the tree looks and behaves. Note that options passed into the the constructor will not be checked to ensure the values are valid. Options will be checked for the "<" and ">" characters. These characters are not permitted to prevent script injection. If options are not passed in, the following default values will be used instead. Options are categorized as impacting the behavior of the tree or the appearance of the tree but can be specified in any order.

[Experiment with options live.](https://observablehq.com/@amogh/dependentree#cell-602)

#### Tree Behavior

| Name                      | Type               | Default Value | Description                                                                                                                                                                                                                                                                                                                                                    |
| ------------------------- | ------------------ | ------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `animationDuration`       | Integer            | `750`         | How many milliseconds a node movement animation lasts.                                                                                                                                                                                                                                                                                                         |
| `maxDepth`                | Integer            | `25`          | See the [Understanding Max Depth Nodes](#understanding-max-depth-nodes) section.                                                                                                                                                                                                                                                                               |
| `enableTooltip`           | Boolean            | `true`        | Enables a tooltip with additional information about each entity when the node is hovered over.                                                                                                                                                                                                                                                                 |
| `enableTooltipKey`        | Boolean            | `true`        | If set to `true`, the key in each key value pair of additional entity information will be displayed. For example: `Title: Queen` as opposed to simply `Queen`.                                                                                                                                                                                                 |
| `modifyEntityName`        | Function           | `null`        | A function which is passed the entity data object and should return a string representing the what the node text will be. The default value for `_name` will be passed in if this variable is `null`. See the [modifyEntityName](#modifyentityname) example below.                                                                                             |
| `textClick`               | Function           | `null`        | A function passed as a callback when users click on the text of the node. This method is passed two arguments, [mouse event](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent) and the [entity data](#terminology) object. If this value is `null`, clicking the text will expand/collapse the node. See the [textClick](#textclick) example below. |
| `maxDepthMessage`         | String or Function | `null`        | Overwrites the max depth message when not `null`. If this option is set to a function, the entity data object will be passed in to the function as an argument; this function should return a string.                                                                                                                                                          |
| `missingEntityMessage`    | String or Function | `null`        | Overwrites the default missing entity message when not `null`. If this option is set to a function, the missing entity's name will be passed in to the function as an argument; this function should return a string.                                                                                                                                          |
| `cyclicDependencyMessage` | String or Function | `null`        | Overwrites the default cyclic dependency loop message when not `null`. If this option is set to a function, the entity data object will be passed in to the function as an argument; this function should return a string.                                                                                                                                     |

#### Tree Appearance

| Name                          | Type    | Default Value                                                                                                    | Description                                                                                                                                                                                                                                                                                                                             |
| ----------------------------- | ------- | ---------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `containerWidthMultiplier`    | Number  | `4`                                                                                                              | The tree takes the default width of the HTML element passed into the selector string and multiplies it by the `containerWidthMultiplier`. It's usually best not to reduce this value. The tree's scale is determined by the width of the container. If you do reduce this value, you may need to adjust the size of the nodes and text. |
| `containerWidthInPx`          | Integer | `null`                                                                                                           | Sets the container width with a constant pixel value. This overrides `containerWidthMultiplier` if the value is not `null`. Take caution with setting this value too low. See the note above.                                                                                                                                           |
| `marginTop`                   | Integer | 60                                                                                                               | Sets the `margin-top` of the tree's wrapping SVG element.                                                                                                                                                                                                                                                                               |
| `marginRight`                 | Integer | 120                                                                                                              | Sets the `margin-right` of the tree's wrapping SVG element.                                                                                                                                                                                                                                                                             |
| `marginBottom`                | Integer | 200                                                                                                              | Sets the `margin-bottom` of the tree's wrapping SVG element.                                                                                                                                                                                                                                                                            |
| `marginLeft`                  | Integer | 120                                                                                                              | Sets the `margin-left` of the tree's wrapping SVG element.                                                                                                                                                                                                                                                                              |
| `parentNodeTextOrientation`   | String  | `'left'`                                                                                                         | Sets the placement of text for a given node that has children. The direction refers to downstream dependencies. This is inverted for upstream.                                                                                                                                                                                          |
| `childNodeTextOrientation`    | String  | `'right'`                                                                                                        | Sets the placement of text for a given node that has no children. The direction refers to downstream dependencies. This is inverted for upstream.                                                                                                                                                                                       |
| `textOffset`                  | Integer | `13`                                                                                                             | How far the placement of the text is from the center of the node in pixels.                                                                                                                                                                                                                                                             |
| `textStyleFont`               | String  | `'12px sans-serif'`                                                                                              | Node text's [shorthand property](https://developer.mozilla.org/en-US/docs/Web/CSS/font).                                                                                                                                                                                                                                                |
| `textStyleColor`              | String  | `'black'`                                                                                                        | Node text's [color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value).                                                                                                                                                                                                                                                      |
| `circleStrokeColor`           | String  | `'steelblue'`                                                                                                    | Outline [color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) of the node circle.                                                                                                                                                                                                                                       |
| `circleStrokeWidth`           | Integer | `3`                                                                                                              | Width of the node circle in pixels.                                                                                                                                                                                                                                                                                                     |
| `circleSize`                  | Integer | `10`                                                                                                             | Size of the node's circle in pixels.                                                                                                                                                                                                                                                                                                    |
| `linkStrokeColor`             | String  | `'#dddddd'`                                                                                                      | [Color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) of the link. A link is the line that connects nodes.                                                                                                                                                                                                              |
| `linkStrokeWidth`             | Integer | `2`                                                                                                              | With of the link in pixels.                                                                                                                                                                                                                                                                                                             |
| `closedNodeCircleColor`       | String  | `'lightsteelblue'`                                                                                               | [Color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) of the circle when the node has children and is closed.                                                                                                                                                                                                           |
| `openNodeCircleColor`         | String  | `'white'`                                                                                                        | [Color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) of the circle when the node has children and is closed or when the node has no children.                                                                                                                                                                          |
| `cyclicNodeColor`             | String  | `'#FF4242'`                                                                                                      | [Color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) of the circle of nodes that have cyclic dependency loops.                                                                                                                                                                                                         |
| `missingNodeColor`            | String  | `'#E8F086'`                                                                                                      | [Color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) of the circle of nodes that were not included in the input entity list but found as a dependency.                                                                                                                                                                 |
| `maxDepthNodeColor`           | String  | `'#A691AE'`                                                                                                      | [Color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) of the circle of nodes that mark the end of the tree that can be displayed.                                                                                                                                                                                       |

CHANGELOG.md
### 1.0.2 / March 10th, 2022
 - Enabled adding DependenTree package with `require()` statements


### 1.0.1 / March 9th, 2022
 - Initialize

src/options.js
/*!
 *
 * Copyright 2022 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

const defaultOptions = {
  // behavior options
  animationDuration: 750,
  maxDepth: 25,
  enableTooltip: true,
  enableTooltipKey: true,
  modifyEntityName: null,
  textClick: null,
  maxDepthMessage: null,
  missingEntityMessage: null,
  cyclicDependencyMessage: null,

  // appearance options
  containerWidthMultiplier: 4,
  containerWidthInPx: null,
  marginTop: 60,
  marginRight: 120,
  marginBottom: 200,
  marginLeft: 120 ,
  parentNodeTextOrientation: 'left',
  childNodeTextOrientation: 'right',
  textOffset: 13,
  textStyleFont: '12px sans-serif',
  textStyleColor: 'black',
  circleStrokeColor: 'steelblue',
  circleStrokeWidth: 3,
  circleSize: 10,
  linkStrokeColor: '#dddddd',
  linkStrokeWidth: 2,
  closedNodeCircleColor: 'lightsteelblue',
  openNodeCircleColor: 'white',
  cyclicNodeColor: '#FF4242',
  missingNodeColor: '#E8F086',
  maxDepthNodeColor: '#A691AE',
  horizontalSpaceBetweenNodes: 180,
  verticalSpaceBetweenNodes: 30,
  wrapNodeName: true,
  splitStr: null,
  tooltipItemStyleObj: {
    'font-family': 'sans-serif',
    'font-size': '12px',
  },
  tooltipColonStr: ': ',
  tooltipKeyStyleObj: { 'font-weight': 'bold' },
  tooltipColonStyleObj: { 'font-weight': 'bold' },
  tooltipValueStyleObj: {},
  tooltipStyleObj: {
    'background-color': 'white',
    border: 'solid',
    'border-width': '1px',
    'border-radius': '5px',
    padding: '10px',
  },
};


export function _setOptions(userOptions) {
  for (const key in userOptions) {
    const opt = userOptions[key];
    if (typeof opt === 'string' && opt.includes('<') && opt.includes('>')) {
      throw new Error('Characters not allowed: "<" and ">" are not permitted as options to prevent script injection.')
    }
  }

  this.options = {
    ...defaultOptions,
    ...userOptions,
  }
}
